File: wechat-mp/scripts/media/delete_media.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
微信公众号素材删除工具
删除已上传的永久素材
"""

import argparse
import json
import os
import requests


def load_config(config_path: str = None):
    """加载配置文件"""
    if config_path is None:
        config_path = os.environ.get('WECHAT_MP_CONFIG') or os.path.expanduser('~/.config/wechat-mp/config.json')

    if not os.path.exists(config_path):
        print(f"配置文件不存在: {config_path}")
        return None

    with open(config_path, 'r', encoding='utf-8') as f:
        return json.load(f)


def get_access_token(appid: str, secret: str) -> str:
    """获取 access_token"""
    url = "https://api.weixin.qq.com/cgi-bin/token"
    params = {
        "grant_type": "client_credential",
        "appid": appid,
        "secret": secret
    }

    response = requests.get(url, params=params, timeout=30)
    data = response.json()

    if "access_token" in data:
        return data["access_token"]
    else:
        print(f"获取 access_token 失败: {data.get('errmsg', 'Unknown error')}")
        return None


def delete_material(media_id: str, access_token: str) -> dict:
    """删除永久素材"""
    url = f"https://api.weixin.qq.com/cgi-bin/material/del_material?access_token={access_token}"

    data = {"media_id": media_id}
    response = requests.post(url, json=data, timeout=30)
    result = response.json()

    if result.get("errcode") == 0:
        return {"success": True}
    else:
        return {"error": result.get("errmsg", "Unknown error")}


def main():
    parser = argparse.ArgumentParser(description='微信公众号素材删除工具')
    parser.add_argument('--media-id', required=True, help='素材 media_id')
    parser.add_argument('--config', help='配置文件路径')
    parser.add_argument('--force', action='store_true', help='跳过确认直接删除')

    args = parser.parse_args()

    # 确认删除
    if not args.force:
        confirm = input(f"确认删除素材 {args.media_id}? (y/N): ")
        if confirm.lower() != 'y':
            print("已取消删除")
            return

    # 加载配置
    config = load_config(args.config)
    if not config:
        return

    wechat_config = config.get('wechat', {})
    appid = wechat_config.get('app_id')
    secret = wechat_config.get('app_secret')

    if not appid or not secret:
        print("配置文件中缺少 app_id 或 app_secret")
        return

    # 获取 access_token
    print("正在获取 access_token...")
    access_token = get_access_token(appid, secret)
    if not access_token:
        return
    print("access_token 获取成功")

    # 删除素材
    print(f"正在删除素材: {args.media_id}")
    result = delete_material(args.media_id, access_token)

    if result.get("success"):
        print("\n删除成功!")
    else:
        print(f"\n删除失败: {result.get('error')}")


if __name__ == '__main__':
    main()

File: wechat-mp/scripts/media/delete_media.mjs
#!/usr/bin/env node
// 微信公众号素材删除工具：删除已上传的永久素材。
// 零第三方依赖，使用 Node 18+ 内置 fetch。

import { readFileSync, existsSync } from 'node:fs';
import { homedir } from 'node:os';
import { createInterface } from 'node:readline/promises';

const TOKEN_URL = 'https://api.weixin.qq.com/cgi-bin/token';

function parseArgs(argv) {
  const args = { force: false };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--media-id') args.mediaId = argv[++i];
    else if (a === '--config') args.config = argv[++i];
    else if (a === '--force') args.force = true;
  }
  return args;
}

function loadConfig(configPath) {
  const path = configPath
    || process.env.WECHAT_MP_CONFIG
    || `${homedir()}/.config/wechat-mp/config.json`;
  if (!existsSync(path)) {
    console.log(`配置文件不存在: ${path}`);
    return null;
  }
  return JSON.parse(readFileSync(path, 'utf-8'));
}

async function getAccessToken(appid, secret) {
  const url = `${TOKEN_URL}?grant_type=client_credential&appid=${encodeURIComponent(appid)}&secret=${encodeURIComponent(secret)}`;
  const data = await (await fetch(url, { signal: AbortSignal.timeout(30000) })).json();
  if (data.access_token) return data.access_token;
  console.log(`获取 access_token 失败: ${data.errmsg || 'Unknown error'}`);
  return null;
}

async function deleteMaterial(mediaId, accessToken) {
  const url = `https://api.weixin.qq.com/cgi-bin/material/del_material?access_token=${accessToken}`;
  const data = await (await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ media_id: mediaId }),
    signal: AbortSignal.timeout(30000),
  })).json();
  if (data.errcode === 0) return { success: true };
  return { error: data.errmsg || 'Unknown error' };
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  if (!args.mediaId) { console.error('缺少必填参数 --media-id'); process.exit(1); }

  if (!args.force) {
    const rl = createInterface({ input: process.stdin, output: process.stdout });
    const confirm = await rl.question(`确认删除素材 ${args.mediaId}? (y/N): `);
    rl.close();
    if (confirm.toLowerCase() !== 'y') { console.log('已取消删除'); return; }
  }

  const config = loadConfig(args.config);
  if (!config) return;

  const { app_id: appid, app_secret: secret } = config.wechat || {};
  if (!appid || !secret) { console.log('配置文件中缺少 app_id 或 app_secret'); return; }

  console.log('正在获取 access_token...');
  const accessToken = await getAccessToken(appid, secret);
  if (!accessToken) return;
  console.log('access_token 获取成功');

  console.log(`正在删除素材: ${args.mediaId}`);
  const result = await deleteMaterial(args.mediaId, accessToken);

  if (result.success) console.log('\n删除成功!');
  else console.log(`\n删除失败: ${result.error}`);
}

main();
